addons/html_editor/static/src/utils/formatting.js

```diff
@@ -118,23 +135,56 @@ export const formatsSpecs = {
                       findNode(
                           closestPath(node),
                           (el) => el.classList?.contains(props.className),
-                          isBlock
+                          (el) => el === closestBlock(node).parentElement
                       ) || closestElement(node, "li")?.classList?.contains(props.className)
                   )
                 : !!findNode(
                       closestPath(node),
                       (el) => FONT_SIZE_CLASSES.find((cls) => el.classList?.contains(cls)),
-                      isBlock
+                      (el) => el === closestBlock(node).parentElement
                   ) ||
                   FONT_SIZE_CLASSES.find((cls) =>
                       closestElement(node, "li")?.classList.contains(cls)
                   ),
-        hasStyle: (node, props) => FONT_SIZE_CLASSES.find((cls) => node.classList.contains(cls)),
+        hasStyle: (node, props) =>
+            [...FONT_SIZE_CLASSES, ...TEXT_STYLE_CLASSES, ...DEFAULT_FONT_SIZE_CLASSES].find(
+                (cls) => node.classList.contains(cls)
+            ),
         addStyle: (node, props) => {
             node.style.removeProperty("font-size");
             node.classList.add(props.className);
         },
-        removeStyle: (node) => removeClass(node, ...FONT_SIZE_CLASSES, ...TEXT_STYLE_CLASSES),
+        removeStyle: (node) => {
+            removeClass(node, ...FONT_SIZE_CLASSES);
+            // Typography classes should be preserved on block elements since
+            // they act as semantic equivalents of <h1>, <h2>, etc., not just
+            // removable styles.
+            if (!isBlock(node)) {
+                removeClass(node, ...TEXT_STYLE_CLASSES, ...DEFAULT_FONT_SIZE_CLASSES);
+            }
+        },
+        addNeutralStyle: function (node, isParentNodeBlockAndCompletelySelected) {
+            const block = closestBlock(node);
+            if (isParentNodeBlockAndCompletelySelected) {
```

Comment:
To avoid having this clunky parameter, maybe we should move this whole `formatSpecs` object to the `FormatPlugin` so that it can compute `isParentNodeBlockAndCompletelySelected` by itself? It's the only plugin using it. Or maybe just change the api a bit to something more palatable, like you would be passing the parent element insted of node when the parent is block and fully selected, and then the second parameter could be `recursive` or something like that. So the idea would be you want to add a neutral style to a node (in some case you'd call the function on the parent node) and then you can say whether you want the neutrality to be applied recursively or not. What do you think?